size-related compilation issues when adding scalar to vector to matrix in model

Issue 1: this fails  if `project` is a vector (but is fine if it is a 3x2 matrix)

```
code<- nimbleCode({
    for(i in 1:4)
    p[1:3,1:2,i] <- B.0[i] + B.1[i] *project[1:3] + B.yday*yday[1:3,1:2]+ B.lag[i]*lag[1:3,1:2,i]
})
m <- nimbleModel(code, data=list(project=rnorm(3),yday=matrix(rnorm(6),3,2), lag=array(rnorm(24),c(3,2,4))))
cm <- compileNimble(m)
```

```
Error: In sizeBinaryCwise: Fixed size mismatch.
 This occurred for: model_B_dot_0[getNodeFunctionIndexedInfo(ARG1_INDEXEDNODEINFO__,1)] + model_B_dot_1[getNodeFunctionIndexedInfo(ARG1_INDEXEDNODEINFO__,1)] * eigenBlock(model_project,1:3) + model_B_dot_yday[1] * eigenBlock(model_yday,1:3,1:2) + model_B_dot_lag[getNodeFunctionIndexedInfo(ARG1_INDEXEDNODEINFO__,1)] * map(name,2,call,list,list)
 This was part of the call:  model_p[1:3, 1:2, getNodeFunctionIndexedInfo(ARG1_INDEXEDNODEINFO__,1)] <<- model_B_dot_0[getNodeFunctionIndexedInfo(ARG1_INDEXEDNODEINFO__,1)] + model_B_dot_1[getNodeFunctionIndexedInfo(ARG1_INDEXEDNODEINFO__,1)] * eigenBlock(model_project,1:3) + model_B_dot_yday[1] * eigenBlock(model_yday,1:3,1:2) + model_B_dot_lag[getNodeFunctionIndexedInfo(ARG1_INDEXEDNODEINFO__,1)] * map(name,2,call,list,list)
```

Issue #2: if the `lag` part of the summation is left out we get a run-time size error:

```
code<- nimbleCode({
    for(i in 1:4)
    p[1:3,1:2,i] <- B.0[i] + B.1[i] *project[1:3] + B.yday*yday[1:3,1:2] # + B.lag[i]*lag[1:3,1:2,i]
})
m <- nimbleModel(code, data=list(project=rnorm(3),yday=matrix(rnorm(6),3,2), lag=array(rnorm(24),c(3,2,4))))
cm <- compileNimble(m)
cm$simulate('p')
```
```
Error in callEnv[[funName]](..., .basePtr = basePtrList[[index]]) : 
  Run-time size error: expected 2 == 1
```

On the other hand, both of these subsets of the summation are fine:

```
code<- nimbleCode({
    for(i in 1:4)
    p[1:3,1:2,i] <- B.0[i] + B.1[i] *project[1:3]
})
m <- nimbleModel(code, data=list(project=rnorm(3),B.0=rnorm(4),B.1=rnorm(4)))
cm <- compileNimble(m)
cm$simulate()

code<- nimbleCode({
    for(i in 1:4)
    p[1:3,1:2,i] <- B.0[i] + B.yday*yday[1:3,1:2]
})
m <- nimbleModel(code, data=list(project=rnorm(3),yday=matrix(rnorm(6),3,2), lag=array(rnorm(24),c(3,2,4))))
cm <- compileNimble(m)
cm$simulate('p')
```

So there seems to be some interaction between arrays, recycling, and what is being added together that fails in some situations but not all.

We can also get compilation errors in related cases:

```
nf <- nimbleFunction(
    run=function(b0 = double()) {
        x <- array(0,c(3,2,4))
        x[1:3,1:2,1:4] <- b0
        returnType(double(3))
        return(x)
    })
cnf <- compileNimble(nf)
```

```
using C++ compiler: ‘g++ (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0’
P_55_rcFun_R_GlobalEnv24.cpp: In function ‘NimArr<3, double> rcFun_R_GlobalEnv24(double)’:
P_55_rcFun_R_GlobalEnv24.cpp:20:13: error: no match for ‘operator=’ (operand types are ‘NimArr<3, double>’ and ‘double’)
   20 | Interm_44 = ARG1_b0_;
      |             ^~~~~~~~
In file included from /accounts/vis/paciorek/R/x86_64-pc-linux-gnu-library-ubuntu-22.04/4.3/nimble/include/nimble/RcppNimbleUtils.h:25,
                 from /accounts/vis/paciorek/R/x86_64-pc-linux-gnu-library-ubuntu-22.04/4.3/nimble/include/nimble/NamedObjects.h:28,
                 from /accounts/vis/paciorek/R/x86_64-pc-linux-gnu-library-ubuntu-22.04/4.3/nimble/include/nimble/EigenTypedefs.h:26,
                 from P_55_rcFun_R_GlobalEnv24.cpp:8:
/accounts/vis/paciorek/R/x86_64-pc-linux-gnu-library-ubuntu-22.04/4.3/nimble/include/nimble/NimArr.h:625:17: note: candidate: ‘NimArr<3, T>& NimArr<3, T>::operator=(const NimArr<3, T>&) [with T = double]’
  625 |   NimArr<3, T> &operator=(const NimArr<3, T> &other) {
      |                 ^~~~~~~~
/accounts/vis/paciorek/R/x86_64-pc-linux-gnu-library-ubuntu-22.04/4.3/nimble/include/nimble/NimArr.h:625:47: note:   no known conversion for argument 1 from ‘double’ to ‘const NimArr<3, double>&’
  625 |   NimArr<3, T> &operator=(const NimArr<3, T> &other) {
      |                           ~~~~~~~~~~~~~~~~~~~~^~~~~
```

--- a/packages/nimble/R/genCpp_sizeProcessing.R
+++ b/packages/nimble/R/genCpp_sizeProcessing.R
@@ -2244,7 +2244,8 @@ sizeAssignAfterRecursing <- function(code, symTab, typeEnv, NoEigenizeMap = FALS
                     if(LHS$nDim == 1) {
                       if(RHS$nDim == 2) {
                         # vector[] <- matrix case
-                            if(is.numeric(RHS$sizeExprs[[1]]) && (RHS$sizeExprs[[1]] == 1)) {
+                        if(is.numeric(RHS$sizeExprs[[1]]) && (RHS$sizeExprs[[1]] == 1)) {
+                           ## simple vector <- matrix[1,] case
                                 newExpr <- insertExprClassLayer(code, 1, 'asRow', type = LHS$type)
                                 newExpr$sizeExprs <- RHS$sizeExprs
                                 newExpr$type <- LHS$type
@@ -2258,26 +2259,33 @@ sizeAssignAfterRecursing <- function(code, symTab, typeEnv, NoEigenizeMap = FALS
                                 }
                             } else {
                               if(isTRUE(getNimbleOption("newAssignSizeChecks"))) {
-                                # print("vector[] <- matrix that is not a simple vector <- matrix row case")
-                                # check that length of LHS matches nrow of RHS
-                                if(!is.numeric(LHS$sizeExprs[[1]]) || !is.numeric(RHS$sizeExprs[[1]])) {
-                                  assertMessage <- paste0("Run-time size error: expected ", deparse(LHS$sizeExprs[[1]]), " == ", deparse(RHS$sizeExprs[[1]]))
-                                  thisAssert <- identityAssert(LHS$sizeExprs[[1]], RHS$sizeExprs[[1]], assertMessage)
+                                # print("vector[] <- matrix that is not a simple vector[] <- matrix[1,]  case")
+                                        # Here the checking is relaxed in the sense that we check the product of the
+                                        # matrix dimensions, which should match the vector dimension.
+
+                                if(!is.numeric(LHS$sizeExprs[[1]]) || !is.numeric(RHS$sizeExprs[[1]]) || !is.numeric(RHS$sizeExprs[[2]])) {
+
+                                  assertMessage <- paste0("Run-time size error: expected ",
+                                                          deparse(LHS$sizeExprs[[1]]), " == (", deparse(RHS$sizeExprs[[1]]),')*(', deparse(RHS$sizeExprs[[2]]))
+                                  thisAssert <- identityAssert(LHS$sizeExprs[[1]],
+                                                               substitute((A)*(B), list(A=RHS$sizeExprs[[1]], B=RHS$sizeExprs[[2]])),
+                                                               assertMessage)
                                   if(!is.null(thisAssert)) assert[[length(assert) + 1]] <- thisAssert
                                 } else {
-                                  if(LHS$sizeExprs[[1]] != RHS$sizeExprs[[1]])
+                                  ## All numeric
+                                  if(LHS$sizeExprs[[1]] != (RHS$sizeExprs[[1]] * RHS$sizeExprs[[2]]))
                                     stop(exprClassProcessingErrorMsg(code,
                                                                      'In sizeAssignAfterRecursing: Fixed size mismatch.'), call. = FALSE)
                                 }
-                                # check that RHS has ncol==1
-                                if(is.numeric(RHS$sizeExprs[[2]]) && (RHS$sizeExprs[[2]] != 1)) {
-                                  stop(exprClassProcessingErrorMsg(code,
-                                                                   'In sizeAssignAfterRecursing: Fixed size mismatch.'), call. = FALSE)
-                                } else {
-                                    assertMessage <- paste0("Run-time size error: expected ", deparse(RHS$sizeExprs[[2]]), " == 1")
-                                    thisAssert <- identityAssert(RHS$sizeExprs[[2]], 1, assertMessage)
-                                    if(!is.null(thisAssert)) assert[[length(assert) + 1]] <- thisAssert
-                                }
+                                ## # check that RHS has ncol==1
+                                ## if(is.numeric(RHS$sizeExprs[[2]]) && (RHS$sizeExprs[[2]] != 1)) {
+                                ##   stop(exprClassProcessingErrorMsg(code,
+                                ##                                    'In sizeAssignAfterRecursing: Fixed size mismatch.'), call. = FALSE)
+                                ## } else {
+                                ##     assertMessage <- paste0("Run-time size error: expected ", deparse(RHS$sizeExprs[[2]]), " == 1")
+                                ##     thisAssert <- identityAssert(RHS$sizeExprs[[2]], 1, assertMessage)
+                                ##     if(!is.null(thisAssert)) assert[[length(assert) + 1]] <- thisAssert
+                                ## }
                               }
                             }
                       } else if(RHS$nDim==1) {
@@ -2301,7 +2309,8 @@ sizeAssignAfterRecursing <- function(code, symTab, typeEnv, NoEigenizeMap = FALS
                         # print('LHS$nDim != 1')
                         if(RHS$nDim == 1) {
                           # print("matrix[] <- vector")
-                           if(is.numeric(LHS$sizeExprs[[1]]) && (LHS$sizeExprs[[1]] == 1)) {
+                          if(is.numeric(LHS$sizeExprs[[1]]) && (LHS$sizeExprs[[1]] == 1)) {
+                            # simple matrix[1,] <- vector case (not even sure if we need this explicitly here?)
                                 newExpr <- insertExprClassLayer(code, 1, 'asRow', type = LHS$type)
                                 newExpr$sizeExprs <- LHS$sizeExprs
                                 newExpr$type <- LHS$type
@@ -2313,28 +2322,34 @@ sizeAssignAfterRecursing <- function(code, symTab, typeEnv, NoEigenizeMap = FALS
                                 } else {
                                     if(RHS$sizeExprs[[1]] != LHS$sizeExprs[[2]]) stop(exprClassProcessingErrorMsg(code, paste0('In sizeAssignAfterRecursing: Fixed size mismatch.')), call. = FALSE)
                                 }
-                            } else {
+                          } else {
+                            # not just simple matrix[1,] <- vector case
                               if(isTRUE(getNimbleOption("newAssignSizeChecks"))) {
                                 # print("matrix[] <- vector that is not a simple matrix row <- vector case")
-                                # check that length of LHS matches nrow of RHS
-                                if(!is.numeric(RHS$sizeExprs[[1]]) || !is.numeric(LHS$sizeExprs[[1]])) {
-                                  assertMessage <- paste0("Run-time size error: expected ", deparse(RHS$sizeExprs[[1]]), " == ", deparse(LHS$sizeExprs[[1]]))
-                                  thisAssert <- identityAssert(RHS$sizeExprs[[1]], LHS$sizeExprs[[1]], assertMessage)
+                                # check only that product of LHS dims equals RHS
+                                if(!is.numeric(RHS$sizeExprs[[1]]) || !is.numeric(LHS$sizeExprs[[1]]) || !is.numeric(LHS$sizeExprs[[2]])) {
+                                  assertMessage <- paste0("Run-time size error: expected (",
+                                                          deparse(LHS$sizeExprs[[1]]), ")*(",
+                                                          deparse(LHS$sizeExprs[[2]]),
+                                                          ") == ", deparse(RHS$sizeExprs[[1]]))
+                                  thisAssert <- identityAssert(RHS$sizeExprs[[1]],
+                                                               substitute((A)*(B), list(A=LHS$sizeExprs[[1]], B=LHS$sizeExprs[[2]])),
+                                                               assertMessage)
                                   if(!is.null(thisAssert)) assert[[length(assert) + 1]] <- thisAssert
                                 } else {
-                                  if(LHS$sizeExprs[[1]] != RHS$sizeExprs[[1]])
+                                  if((LHS$sizeExprs[[1]]*LHS$sizeExprs[[2]]) != RHS$sizeExprs[[1]])
                                     stop(exprClassProcessingErrorMsg(code,
                                                                      'In sizeAssignAfterRecursing: Fixed size mismatch.'), call. = FALSE)
                                 }
-                                # check that LHS has ncol==1
-                                if(is.numeric(LHS$sizeExprs[[2]]) && (LHS$sizeExprs[[2]] != 1)) {
-                                  stop(exprClassProcessingErrorMsg(code,
-                                                                   'In sizeAssignAfterRecursing: Fixed size mismatch.'), call. = FALSE)
-                                } else {
-                                    assertMessage <- paste0("Run-time size error: expected ", deparse(LHS$sizeExprs[[2]]), " == 1")
-                                    thisAssert <- identityAssert(LHS$sizeExprs[[2]], 1, assertMessage)
-                                    if(!is.null(thisAssert)) assert[[length(assert) + 1]] <- thisAssert
-                                }
+                                ## # check that LHS has ncol==1
+                                ## if(is.numeric(LHS$sizeExprs[[2]]) && (LHS$sizeExprs[[2]] != 1)) {
+                                ##   stop(exprClassProcessingErrorMsg(code,
+                                ##                                    'In sizeAssignAfterRecursing: Fixed size mismatch.'), call. = FALSE)
+                                ## } else {
+                                ##     assertMessage <- paste0("Run-time size error: expected ", deparse(LHS$sizeExprs[[2]]), " == 1")
+                                ##     thisAssert <- identityAssert(LHS$sizeExprs[[2]], 1, assertMessage)
+                                ##     if(!is.null(thisAssert)) assert[[length(assert) + 1]] <- thisAssert
+                                ## }
                               }
                             }
                         } else if(RHS$nDim==2) {

--- a/packages/nimble/R/genCpp_sizeProcessing.R
+++ b/packages/nimble/R/genCpp_sizeProcessing.R
@@ -2199,7 +2199,7 @@ sizeAssignAfterRecursing <- function(code, symTab, typeEnv, NoEigenizeMap = FALS
     ## Note this can use LHS$name for RHSsizeExprs when returning from a nimbleFunction on RHS.  But this is probably not needed any more.
     if(any(unlist(lapply(RHSsizeExprs, is.null)))) RHSsizeExprs <- makeSizeExpressions(rep(NA, RHSnDim), LHS$name) ## reset sizeExprs for the LHS var. re-using RHSsizeExprs for LHS.  This would only be valid if it is a nimbleFunction returning something on the RHS.  For assignment to be executed in Eigen, the RHS sizes MUST be known
 
-     if(!getNimbleOption('experimentalNewSizeProcessing')) {
+  if(!getNimbleOption('experimentalNewSizeProcessing')) {
                 
     if(LHS$toEigenize == 'yes') {
         code$toEigenize <- 'yes'
@@ -2240,29 +2240,123 @@ sizeAssignAfterRecursing <- function(code, symTab, typeEnv, NoEigenizeMap = FALS
                     assert <- list(assert)
                 } else { ## We have an indexed LHS of an eigenizable expression
                     ## need special handling if it is a row assignment like x[i,] <- ...
-                    ## also need to generate size assertions                    
+                  ## also need to generate size assertions
                     if(LHS$nDim == 1) {
-                        if(RHS$nDim == 2) {
-                            if(is.numeric(RHS$sizeExprs[[1]])) {
-                                if(RHS$sizeExprs[[1]] == 1) {
-                                    newExpr <- insertExprClassLayer(code, 1, 'asRow', type = LHS$type)
-                                    newExpr$sizeExprs <- RHS$sizeExprs 
-                                    newExpr$type <- LHS$type
-                                    newExpr$nDim <- RHS$nDim
-                                    if(!is.numeric(LHS$sizeExprs[[1]]) || !is.numeric(RHS$sizeExprs[[2]])) {
-                                        assertMessage <- paste0("Run-time size error: expected ", deparse(LHS$sizeExprs[[1]]), " == ", deparse(RHS$sizeExprs[[2]]))
-                                        thisAssert <- identityAssert(LHS$sizeExprs[[1]], RHS$sizeExprs[[2]], assertMessage)
-                                        if(!is.null(thisAssert)) assert[[length(assert) + 1]] <- thisAssert 
-                                    } else {
-                                        if(LHS$sizeExprs[[1]] != RHS$sizeExprs[[2]]) stop(exprClassProcessingErrorMsg(code, paste0('In sizeAssignAfterRecursing: Fixed size mismatch.')), call. = FALSE)
-                                    }
+                      if(RHS$nDim == 2) {
+                        # vector[] <- matrix case
+                            if(is.numeric(RHS$sizeExprs[[1]]) && (RHS$sizeExprs[[1]] == 1)) {
+                                newExpr <- insertExprClassLayer(code, 1, 'asRow', type = LHS$type)
+                                newExpr$sizeExprs <- RHS$sizeExprs
+                                newExpr$type <- LHS$type
+                                newExpr$nDim <- RHS$nDim
+                                if(!is.numeric(LHS$sizeExprs[[1]]) || !is.numeric(RHS$sizeExprs[[2]])) {
+                                    assertMessage <- paste0("Run-time size error: expected ", deparse(LHS$sizeExprs[[1]]), " == ", deparse(RHS$sizeExprs[[2]]))
+                                    thisAssert <- identityAssert(LHS$sizeExprs[[1]], RHS$sizeExprs[[2]], assertMessage)
+                                    if(!is.null(thisAssert)) assert[[length(assert) + 1]] <- thisAssert
+                                } else {
+                                    if(LHS$sizeExprs[[1]] != RHS$sizeExprs[[2]]) stop(exprClassProcessingErrorMsg(code, paste0('In sizeAssignAfterRecursing: Fixed size mismatch.')), call. = FALSE)
                                 }
+                            } else {
+                              if(isTRUE(getNimbleOption("newAssignSizeChecks"))) {
+                                # print("vector[] <- matrix that is not a simple vector <- matrix row case")
+                                # check that length of LHS matches nrow of RHS
+                                if(!is.numeric(LHS$sizeExprs[[1]]) || !is.numeric(RHS$sizeExprs[[1]])) {
+                                  assertMessage <- paste0("Run-time size error: expected ", deparse(LHS$sizeExprs[[1]]), " == ", deparse(RHS$sizeExprs[[1]]))
+                                  thisAssert <- identityAssert(LHS$sizeExprs[[1]], RHS$sizeExprs[[1]], assertMessage)
+                                  if(!is.null(thisAssert)) assert[[length(assert) + 1]] <- thisAssert
+                                } else {
+                                  if(LHS$sizeExprs[[1]] != RHS$sizeExprs[[1]])
+                                    stop(exprClassProcessingErrorMsg(code,
+                                                                     'In sizeAssignAfterRecursing: Fixed size mismatch.'), call. = FALSE)
+                                }
+                                # check that RHS has ncol==1
+                                if(is.numeric(RHS$sizeExprs[[2]]) && (RHS$sizeExprs[[2]] != 1)) {
+                                  stop(exprClassProcessingErrorMsg(code,
+                                                                   'In sizeAssignAfterRecursing: Fixed size mismatch.'), call. = FALSE)
+                                } else {
+                                    assertMessage <- paste0("Run-time size error: expected ", deparse(RHS$sizeExprs[[2]]), " == 1")
+                                    thisAssert <- identityAssert(RHS$sizeExprs[[2]], 1, assertMessage)
+                                    if(!is.null(thisAssert)) assert[[length(assert) + 1]] <- thisAssert
+                                }
+                              }
                             }
+                      } else if(RHS$nDim==1) {
+                        # vector[] <- vector case
+                          if(isTRUE(getNimbleOption("newAssignSizeChecks"))) {
+                            # print('RHS$nDim != 2')
+                            if(!is.numeric(LHS$sizeExprs[[1]]) || !is.numeric(RHS$sizeExprs[[1]])) {
+                              assertMessage <- paste0("Run-time size error: expected ", deparse(LHS$sizeExprs[[1]]), " == ", deparse(RHS$sizeExprs[[1]]))
+                              thisAssert <- identityAssert(LHS$sizeExprs[[1]], RHS$sizeExprs[[1]], assertMessage)
+                              if(!is.null(thisAssert)) assert[[length(assert) + 1]] <- thisAssert
+                            } else {
+                              if(LHS$sizeExprs[[1]] != RHS$sizeExprs[[1]])
+                                stop(exprClassProcessingErrorMsg(code,
+                                                                 'In sizeAssignAfterRecursing: Fixed size mismatch.'), call. = FALSE)
+                            }
+                          }
                         }
-                    }
-                }
-            }
-        }
+                    } else if(LHS$nDim==2) {
+                      # matrix[] <- [matrix|vector]
+                      if(isTRUE(getNimbleOption("newAssignSizeChecks"))) {
+                        # print('LHS$nDim != 1')
+                        if(RHS$nDim == 1) {
+                          # print("matrix[] <- vector")
+                           if(is.numeric(LHS$sizeExprs[[1]]) && (LHS$sizeExprs[[1]] == 1)) {
+                                newExpr <- insertExprClassLayer(code, 1, 'asRow', type = LHS$type)
+                                newExpr$sizeExprs <- LHS$sizeExprs
+                                newExpr$type <- LHS$type
+                                newExpr$nDim <- LHS$nDim
+                                if(!is.numeric(RHS$sizeExprs[[1]]) || !is.numeric(LHS$sizeExprs[[2]])) {
+                                    assertMessage <- paste0("Run-time size error: expected ", deparse(RHS$sizeExprs[[1]]), " == ", deparse(LHS$sizeExprs[[2]]))
+                                    thisAssert <- identityAssert(RHS$sizeExprs[[1]], LHS$sizeExprs[[2]], assertMessage)
+                                    if(!is.null(thisAssert)) assert[[length(assert) + 1]] <- thisAssert
+                                } else {
+                                    if(RHS$sizeExprs[[1]] != LHS$sizeExprs[[2]]) stop(exprClassProcessingErrorMsg(code, paste0('In sizeAssignAfterRecursing: Fixed size mismatch.')), call. = FALSE)
+                                }
+                            } else {
+                              if(isTRUE(getNimbleOption("newAssignSizeChecks"))) {
+                                # print("matrix[] <- vector that is not a simple matrix row <- vector case")
+                                # check that length of LHS matches nrow of RHS
+                                if(!is.numeric(RHS$sizeExprs[[1]]) || !is.numeric(LHS$sizeExprs[[1]])) {
+                                  assertMessage <- paste0("Run-time size error: expected ", deparse(RHS$sizeExprs[[1]]), " == ", deparse(LHS$sizeExprs[[1]]))
+                                  thisAssert <- identityAssert(RHS$sizeExprs[[1]], LHS$sizeExprs[[1]], assertMessage)
+                                  if(!is.null(thisAssert)) assert[[length(assert) + 1]] <- thisAssert
+                                } else {
+                                  if(LHS$sizeExprs[[1]] != RHS$sizeExprs[[1]])
+                                    stop(exprClassProcessingErrorMsg(code,
+                                                                     'In sizeAssignAfterRecursing: Fixed size mismatch.'), call. = FALSE)
+                                }
+                                # check that LHS has ncol==1
+                                if(is.numeric(LHS$sizeExprs[[2]]) && (LHS$sizeExprs[[2]] != 1)) {
+                                  stop(exprClassProcessingErrorMsg(code,
+                                                                   'In sizeAssignAfterRecursing: Fixed size mismatch.'), call. = FALSE)
+                                } else {
+                                    assertMessage <- paste0("Run-time size error: expected ", deparse(LHS$sizeExprs[[2]]), " == 1")
+                                    thisAssert <- identityAssert(LHS$sizeExprs[[2]], 1, assertMessage)
+                                    if(!is.null(thisAssert)) assert[[length(assert) + 1]] <- thisAssert
+                                }
+                              }
+                            }
+                        } else if(RHS$nDim==2) {
+                          # print("matrix[] <- matrix case")
+                          for(iDim in 1:2) {
+                            if(!is.numeric(LHS$sizeExprs[[iDim]]) || !is.numeric(RHS$sizeExprs[[iDim]])) {
+                              assertMessage <- paste0("Run-time size error: expected ", deparse(LHS$sizeExprs[[iDim]]), " == ", deparse(RHS$sizeExprs[[iDim]]))
+                              thisAssert <- identityAssert(LHS$sizeExprs[[iDim]], RHS$sizeExprs[[iDim]], assertMessage)
+                              if(!is.null(thisAssert)) assert[[length(assert) + 1]] <- thisAssert
+                            } else {
+                              if(LHS$sizeExprs[[iDim]] != RHS$sizeExprs[[iDim]])
+                                stop(exprClassProcessingErrorMsg(code,
+                                                                 'In sizeAssignAfterRecursing: Fixed size mismatch.'), call. = FALSE)
+                            }
+                          }
+                        }
+                      }
+                    } # end LHS$nDim == 2
+                } # end else (NOT if(LHS$isName || LHS$name == "nfVar") )
+            } #end if(!(RHS$name %in% setSizeNotNeededOperators))
+        } # end if(RHSnDim > 0)
+      # end if(code$toEigenize == 'yes')
     } else {
         if(inherits(RHS, 'exprClass')) {
             ## If we have A <- map(B, ...), we need to generate a setMap for the RHS, which will be done by sizeInsertIntermediate 
@@ -2273,7 +2367,7 @@ sizeAssignAfterRecursing <- function(code, symTab, typeEnv, NoEigenizeMap = FALS
             if(LHS$name == 'map') assert <- c(assert, sizeInsertIntermediate(code, 1, symTab, typeEnv) )
         }
     }
-     } ##experimentalNewSizeProcessing
+  } ##end if(experimentalNewSizeProcessing)
     if(!(LHS$name %in% c('eigenBlock', 'diagonal', 'coeffSetter', 'nimNonseqIndexedd', 'nimNonseqIndexedi','nimNonseqIndexedb'))) {
         ## should already be annotated if it is an indexed assignment.
         ## It should be harmless to re-annotated EXCEPT in case like out[1:5] <- scalar

--- a/packages/nimble/R/options.R
+++ b/packages/nimble/R/options.R
@@ -7,6 +7,7 @@ nimbleUserNamespace <- as.environment(list(sessionSpecificDll = NULL))
 # These options are for development use at this point.
 .nimbleOptions <- as.environment(
     list(
+        newAssignSizeChecks = TRUE, # If TRUE, do additional checking when [matrix|vector] <- [matrix|vector]. Eventually if successful this option can be removed. It is a failsafe to disable new checks if they give spurious traps.
         allowNFobjInModel = TRUE, # If TRUE, allow use of nimbleFunctions with setup code as model dist or fxn.
         useCppADoptimize = TRUE,
         useADcholAtomic = TRUE, # If TRUE, use nimble's CppAD atomic for cholesky decomposition

